### PR TITLE
Added a gtest for the kernel communications.

### PR DESCRIPTION
**Pull request by mofarrell**
This test does not evaluate the functionality of the kernel
communication unless the KERNEL_TEST flag was set during the build.
The test will not succeed unless the tests are being run as root.

### CODE REVIEW

File: kernel/src/osquery.cpp [hunk 1]
```diff
@@ -247,13 +247,13 @@ static int osquery_ioctl(dev_t dev, u_long cmd, caddr_t data,
...
   }
-#endif // DEBUG
+#endif // KERNEL_TEST
 
   lck_mtx_lock(osquery.mtx);
 
   int err = 0;
   osquery_subscription_args_t *sub;
-  osquery_buf_update_args_t *update;
+  osquery_buf_sync_args_t *sync;
```
Comment (theopolis) [inline kernel/src/osquery.cpp]: Is it possible to initialize these `= 0;`'?

File: kernel/include/feeds.h [hunk 2]
```diff
@@ -25,15 +30,15 @@ extern "C" {
 typedef enum osquery_event {
   END_OF_BUFFER_EVENT = 0,  // Null event used to signal the end of the buffer.
 
-#ifdef DEBUG
+#ifdef KERNEL_TEST
```
Comment (theopolis) [inline kernel/include/feeds.h]: Would you mind adding some inline docs/comments for what `KERNEL_TEST` is used for?

File: osquery/events/kernel/tests/kernel_communication_tests.cpp [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+#include <gtest/gtest.h>
+
+#include "osquery/dispatcher/dispatcher.h"
+#include "osquery/events/kernel/circular_queue_user.h"
+
+namespace osquery {
+
+class KernelCommunicationTests : public testing::Test {};
+
+class ProducerRunnable : public InternalRunnable {
```
Comment (marpaia) [inline osquery/events/kernel/tests/kernel_communication_tests.cpp]: nit: KernelProducerRunnable?

**Comment by mofarrell** [comment]
ok to test

File: osquery/events/kernel/tests/kernel_communication_tests.cpp [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+#endif  // KERNEL_TEST
+      }
+
+      close(fd);
+    }
+  }
+};
+
+TEST_F(KernelCommunicationTests, test_communication) {
+#ifdef KERNEL_TEST
```
Comment (theopolis) [inline osquery/events/kernel/tests/kernel_communication_tests.cpp]: Maybe we drop the whole test macro into the `#ifdef`?

File: osquery/events/kernel/tests/kernel_communication_tests.cpp [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  auto& dispatcher = Dispatcher::instance();
+
+  for (int c = 0; c < num_threads; ++c) {
+    dispatcher.add(
+        OSQUERY_THRIFT_POINTER::make_shared<KernelProducerRunnable>(
+          events_per_thread, c % 2));
+  }
+
+  osquery_event_t event;
+  void *event_buf = NULL;
```
Comment (theopolis) [inline osquery/events/kernel/tests/kernel_communication_tests.cpp]: nit, In C++ you can set this to `nullptr`.

File: osquery/events/kernel/tests/kernel_communication_tests.cpp [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+#include "osquery/dispatcher/dispatcher.h"
+#include "osquery/events/kernel/circular_queue_user.h"
+
+namespace osquery {
+
+class KernelCommunicationTests : public testing::Test {};
+
+class KernelProducerRunnable : public InternalRunnable {
+ public:
+  int events_to_produce_;
```
Comment (theopolis) [inline osquery/events/kernel/tests/kernel_communication_tests.cpp]: nit, these two member variables can be private.

File: osquery/events/kernel/tests/kernel_communication_tests.cpp [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+ *  Copyright (c) 2014, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <boost/make_shared.hpp>
```
Comment (theopolis) [inline osquery/events/kernel/tests/kernel_communication_tests.cpp]: nit, mind reordering the includes to drop boost below our c&c++ includes?

File: kernel/src/osquery.cpp [hunk 8]
```diff
@@ -359,7 +360,11 @@ kern_return_t OSQueryStart(kmod_info_t *ki, void *d) {
 kern_return_t OSQueryStop(kmod_info_t *ki, void *d) {
   dbg_printf("OSQuery kernel module stoping!\n");
```
Comment (theopolis) [inline kernel/src/osquery.cpp]: Let's drop a "osquery" in the macro printf.

**Comment by mofarrell** [comment]
ok to test

File: kernel/src/osquery.cpp (hunk 8 above)
Comment (marpaia) [inline kernel/src/osquery.cpp]:
```
$ python
Python 2.6.6 (r266:84292, Jan 22 2014, 09:42:36)
[GCC 4.4.7 20120313 (Red Hat 4.4.7-4)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> "OSQuery".lower()
'osquery'
>>>
```

**Pull request merged**